working/0546-patterns/patterns-feature-specification.md

```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
+  switch (color) {
+    case Color.red:
+      return Color.orange;
+    case Color.orange:
+      return Color.yellow;
+    case Color.yellow:
+      return Color.green;
+    case Color.green:
+      return Color.blue;
+    case Color.blue:
+      return Color.purple;
+    case Color.purple:
+      return Color.red;
+  }
 }
 ```
 
-**TODO: Add a `...` syntax to allow ignoring positional fields?**
+And turns it into:
 
-#### Variable matcher
+```dart
+Color shiftHue(Color color) {
+  return switch (color) {
+    case Color.red => Color.orange;
+    case Color.orange => Color.yellow;
+    case Color.yellow => Color.green;
+    case Color.green => Color.blue;
+    case Color.blue => Color.purple;
+    case Color.purple => Color.red;
+  };
+}
+```
 
-A variable matcher lets a matching pattern also perform variable binding.
+The grammar is:
 
 ```
-variableMatcher ::= ( 'final' | 'var' | 'final'? typeWithBinder ) identifier
-```
+primary               ::= // Existing productions...
+                        | switchExpression
 
-By using variable matchers as subpatterns of a larger matched pattern, a single
-composite pattern can validate some condition and then bind one or more
-variables only when that condition holds.
-
-A variable pattern can also have a type annotation in order to only match values
-of the specified type.
+switchExpression      ::= 'switch' '(' expression ')' '{'
+                          switchExpressionCase* defaultExpressionCase? '}'
+switchExpressionCase  ::= caseHead '=>' expression ';'
+defaultExpressionCase ::= 'default' '=>' expression ';'
+```
 
-#### Extractor matcher
+Slotting into `primary` means it can be used anywhere any expression can appear,
+even as operands to unary and binary operators. Many of these uses are ugly, but
+not any more problematic than using a collection literal in the same context
+since a `switch` expression is always delimited by a `switch` and `}`.
 
-An extractor combines a type test and record destructuring. It matches if the
-object has the named type. If so, it then uses the following record pattern to
-destructure fields on the value as that type. This pattern is particularly
-useful for writing code in an algebraic datatype style. For example:
+Making it high precedence allows useful patterns like:
 
 ```dart
-class Rect {
-  final double width, height;
-
-  Rect(this.width, this.height);
-}
+await switch (n) {
+  case 1 => aFuture;
+  case 2 => anotherFuture;
+};
 
-display(Object obj) {
-  switch (obj) {
-    case Rect(width:, height:):
-      print('Rect $width x $height');
-    case _:
-      print(obj);
-  }
-}
+var x = switch (n) {
+  case 1 => obj;
+  case 2 => another;
+}.someMethod();
 ```
 
-You can also use an extractor to both match an enum value and destructure
-fields from it:
+Over half of the switch cases in a large corpus of packages contain either a
+single return statement or an assignment followed by a break so there is some
+evidence this will be useful.
```

Comment:
Let's consider, once more, whether we want `=>` instead of `:` here.

It definitely makes the expression switch stand out as different from the statement switch.
Just want to think a little harder about whether that's a good thing or a bad thing.

Switching from one to the other requires changing every `:` to `=>` ... but it's probably `: return`, so ... maybe it makes sense.
But `=>` usually means `return`, and here it doesn't.
... conflicted...!

Reply:
Yeah, I'm on the fence. There's an open issue discussing this, so I think we can leave it as-is in this PR.